File: Leetcode Weekly Contest Solution/1738. Find Kth Largest XOR Coordinate Value.py
class Solution:
    def kthLargestValue(self, matrix: List[List[int]], k: int) -> int:
        
        
        m, n = len(matrix), len(matrix[0])
        
        dp = [[0 for j in range(n+1)] for i in range(m+1)]
        
        
        heap = []
        memo = 0
        for i in range(1, m+1):
            for j in range(1, n+1):
                
                dp[i][j] = matrix[i-1][j-1] ^ dp[i-1][j] ^ dp[i][j-1] ^ dp[i-1][j-1]
                
                if len(heap) < k:
                    heapq.heappush(heap, dp[i][j])
                    continue
                
                heapq.heappushpop(heap, dp[i][j])
        

        return heap[0]

File: Leetcode Weekly Contest Solution/1738. Find Kth Largest XOR Coordinate Value.cpp

const int N = 1000001;
int q[N];

class Solution {
public:
    int kthLargestValue(vector<vector<int>>& w, int k) {
        
        int n = w.size(), m = w[0].size();
        int cnt = 0;
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                
                if (i) w[i][j] ^= w[i-1][j];
                if (j) w[i][j] ^= w[i][j-1];
                if (i && j) w[i][j] ^= w[i-1][j-1];
                q[cnt++] = w[i][j];
            }
        }
            
        k = cnt - k;
        nth_element(q, q+k, q+cnt);
        return q[k];
    }
};